protos/core.proto

```diff
@@ -260,127 +250,422 @@ message CompiledGraph {
+    // Target is the unique action identifier.
+    Target target = 1;
+
+    // Canonical Target is the unique action identifier, with no target
+    // overrides applied.
+    // This option cannot be manually specified in action configs.
+    Target canonical_target = 2;
+
+    // Targets (non-canonical) of actions that this action is dependent on.
+    repeated Target dependency_targets = 3;
+
+    // Path to the source file that the contents of the action is loaded from.
+    string source = 4;
+
+    // Path that the config of the action is loaded from.
+    // This option cannot be manually specified in action configs.
+    string config_path = 5;
+
+    // A list of user-defined tags with which the action should be labeled.
+    repeated string tags = 6;
+
+    // Description of the notebook.
+    string description = 7;
+
+    // TODO(ekrekr): add a notebook runtime field definition.
+  }
+
+  message Actions {
+    message Action {
+      oneof action {
+        Table table = 1;
+        View view = 2;
+        IncrementalTable incremental_table = 3;
+        Assertion assertion = 4;
+        Operation operation = 5;
+        Declaration declaration = 6;
+        Notebook notebook = 7;
+      }
+    }
+    repeated Action action = 1;
+  }
+
+  Actions actions = 1;
+
+  message WorkflowSettings {
+    string dataform_core_version = 1;
+
+    string default_database = 2;
+    string default_schema = 3;
+    string default_location = 4;
+
+    string assertion_schema = 5;
+
+    map<string, string> vars = 6;
+
+    string database_suffix = 7;
+    string schema_suffix = 8;
+    string table_prefix = 9;
+  }
+
+  WorkflowSettings workflow_settings = 2;
+
+  GraphErrors graph_errors = 3;
+
+  string dataform_core_version = 4;
+
+  repeated Target targets = 5;
 }
 
-// Configuration options for BigQuery table opertions.
-message OperationConfig {
-  // If set to true, this action will not be executed. However, the action may
-  // still be depended upon.
-  bool disabled = 1;
+message CoreExecutionRequest {
+  oneof request {
+    CompileExecutionRequest compile = 1;
+  }
+}
 
-  // Declares that this `operations` action creates a dataset which should be
-  // referenceable using the `ref` function.
-  bool has_output = 2;
+// Core execution response contains multiple compiled graphs, that are
+// versioned, because it facilitates consumers supporting major protobuf changes
+// within the same proto file.
+message CoreExecutionResponse {
+  oneof response {
+    CompileExecutionResponse compile = 1 [deprecated = true];
+    CompileExecutionResponseV3 compile_v3 = 2;
+  }
 }
 
-// Configuration options for BigQuery table declarations.
-// TODO(ekrekr): remove this empty message, if explicit action type without a
-// `type` field proves to be impossible / ugly.
-message DeclarationConfig {}
+message CompileExecutionRequest {
+  CompileConfig compile_config = 1;
+}
+message CompileExecutionResponse {
+  CompiledGraph compiled_graph = 1;
```

Comment:
nit: I'd say the oneof goes here. It's still a "compile execution response" in either case